# Car Listings Preprocessing

## Goal:
##### 🔻 Process the cars DataFrame to prepare it for analysis.

## Steps:
    ▼ General understanding of data ' Explore the dataset '
    ▼ Handle noise data for each column
    ▼ Remove unnecessary columns
    ▼ Handle missing data
    ▼ save new df 

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [2]:
pd.set_option("display.max_columns", None)
df = pd.read_csv('../csv_files/draft_df.csv')
df

Make  Year   Price  Color Fuel Type Car Origin Car License  \
0           كيا برايد  2008   44000   أسود     بنزين      خصوصي    فلسطينية   
1        هونداي اكسنت  2013   58000    فضي     بنزين      خصوصي    فلسطينية   
2            بيجو 508  2013   40000   أسود     بنزين      خصوصي    فلسطينية   
3      سكودا اوكتافيا  2010   70000   أزرق      ديزل      خصوصي    فلسطينية   
4      سوبارو امبريزا  1999   33000  رصاصي     بنزين      خصوصي    فلسطينية   
...               ...   ...     ...    ...       ...        ...         ...   
6859    فيات فيات 132  1973   27000   أحمر     بنزين      خصوصي    فلسطينية   
6860  هونداي سانتافيه  2016  117000   أسود      ديزل      خصوصي    فلسطينية   
6861     هونداي اكسنت  2017   67000    فضي     بنزين      خصوصي    فلسطينية   
6862      سكودا فابيا  2015   58000   أبيض     بنزين      خصوصي    فلسطينية   
6863        سيت ابيزا  2016   65000   سكني     بنزين      خصوصي    فلسطينية   

     Transmission Type     Glass  Engine Power Mileage   Payment Method  \
0            اوتوماتيك  الكتروني          1600  ١٨٣٠٠٠  إمكانية التقسيط   
1            اوتوماتيك  الكتروني          1400   42000         نقدا فقط   
2        نصف اوتوماتيك  الكتروني          2000   85000  إمكانية التقسيط   
3                 عادي  الكتروني          1900   30000  إمكانية التقسيط   
4                 عادي  الكتروني          1600     NaN         نقدا فقط   
...                ...       ...           ...     ...              ...   
6859              عادي      يدوي          1800   50000         نقدا فقط   
6860         اوتوماتيك  الكتروني          2000     140         نقدا فقط   
6861         اوتوماتيك  الكتروني          1400      40         نقدا فقط   
6862         اوتوماتيك  الكتروني          1200  60 الف         نقدا فقط   
6863              عادي  الكتروني          1400   82الف  إمكانية التقسيط   

             Available Previous Owners  \
0            للبيع فقط         يد اولى   
1     للبيع أو التبديل         يد اولى   
2            للبيع فقط        يد ثانيه   
3     للبيع أو التبديل               3   
4            للبيع فقط             NaN   
...                ...             ...   
6859         للبيع فقط             NaN   
6860         للبيع فقط             NaN   
6861         للبيع فقط               1   
6862  للبيع أو التبديل               2   
6863         للبيع فقط               1   

                                              Additions  Air Conditioner  \
0     مُكيّف\nإغلاق مركزي\nجهاز إنذار\nمسجل CD\nفرش ...                1   
1     مُكيّف\nإغلاق مركزي\nجهاز إنذار\nجنطات مغنيسيو...                1   
2     مُكيّف\nإغلاق مركزي\nجهاز إنذار\nمسجل CD\nفتحة...                1   
3     مُكيّف\nإغلاق مركزي\nجهاز إنذار\nمسجل CD\nجنطا...                1   
4     مُكيّف\nإغلاق مركزي\nجهاز إنذار\nمسجل CD\nجنطا...                1   
...                                                 ...              ...   
6859                                   مسجل CD\nفرش جلد                0   
6860  مُكيّف\nإغلاق مركزي\nجهاز إنذار\nفتحة سقف\nجنط...                1   
6861  مُكيّف\nإغلاق مركزي\nجهاز إنذار\nمسجل CD\nجنطا...                1   
6862  مُكيّف\nإغلاق مركزي\nجهاز إنذار\nمسجل CD\nجنطا...                1   
6863  مُكيّف\nإغلاق مركزي\nجهاز إنذار\nجنطات مغنيسيو...                1   

      Central Locking  Alarm System  CD Player  Sunroof  Alloy Wheels  \
0                   1             1          1        0             0   
1                   1             1          0        0             1   
2                   1             1          1        1             1   
3                   1             1          1        0             1   
4                   1             1          1        0             1   
...               ...           ...        ...      ...           ...   
6859                0             0          1        0             0   
6860                1             1          0        1             1   
6861                1             1          1        0             1   
6862               

#### General understanding of data

In [3]:
df.shape

(6864, 25)

In [4]:
df.describe()

Year          Price  Engine Power  Air Conditioner  \
count  6864.000000    6864.000000   6864.000000      6864.000000   
mean   2008.616259   66933.643211   1740.494755         0.839307   
std       8.690874   57022.973482    589.470714         0.367275   
min    1967.000000    1111.000000    106.000000         0.000000   
25%    2005.000000   36000.000000   1400.000000         1.000000   
50%    2011.000000   55000.000000   1600.000000         1.000000   
75%    2015.000000   83000.000000   2000.000000         1.000000   
max    2022.000000  999999.000000   9000.000000         1.000000   

       Central Locking  Alarm System    CD Player      Sunroof  Alloy Wheels  \
count      6864.000000   6864.000000  6864.000000  6864.000000   6864.000000   
mean          0.813811      0.673223     0.843386     0.264423      0.731206   
std           0.389287      0.469069     0.363463     0.441058      0.443365   
min           0.000000      0.000000     0.000000     0.000000      0.000000   
25%           1.000000      0.000000     1.000000     0.000000      0.000000   
50%           1.000000      1.000000     1.000000     0.000000      1.000000   
75%           1.000000      1.000000     1.000000     1.000000      1.000000   
max           1.000000      1.000000     1.000000     1.000000      1.000000   

       Leather Seats       Airbag  
count    6864.000000  6864.000000  
mean        0.522290     0.771125  
std         0.499539     0.420139  
min         0.000000     0.000000  
25%         0.000000     1.000000  
50%         1.000000     1.000000  
75%         1.000000     1.000000  
max         1.000000     1.000000

In [5]:
df.head(20)

Make  Year   Price        Color Fuel Type Car Origin  \
0          كيا برايد  2008   44000         أسود     بنزين      خصوصي   
1       هونداي اكسنت  2013   58000          فضي     بنزين      خصوصي   
2           بيجو 508  2013   40000         أسود     بنزين      خصوصي   
3     سكودا اوكتافيا  2010   70000         أزرق      ديزل      خصوصي   
4     سوبارو امبريزا  1999   33000        رصاصي     بنزين      خصوصي   
5           سيت ليون  2017   78000  أسود ميتالك     بنزين      خصوصي   
6         مرسيدس 220  1999   50000         خمري      ديزل      عمومي   
7            اودي 80  1985   18000      برتقالي     بنزين      خصوصي   
8       هونداي كوبيه  1999   30000         أحمر     بنزين      خصوصي   
9    فولكسفاجن كبينة  1979   60000         أصفر     بنزين      خصوصي   
10    فولكسفاجن جولف  2010   70000  أسود ميتالك     بنزين      خصوصي   
11         Dodge رام  2015  125000         أبيض      ديزل      خصوصي   
12         سيت قرطبة  2001   25000         أزرق      ديزل      عمومي   
13       هونداي جيتس  2007   32000         ذهبي     بنزين      خصوصي   
14    سكودا اوكتافيا  2005   42000         أبيض      ديزل      عمومي   
15          بيجو 208  2016   39500         أبيض     بنزين      خصوصي   
16       سكودا رابيد  2017   78000         سكني      ديزل      خصوصي   
17     سكودا فيليسيا  2000   25000         أبيض      ديزل      خصوصي   
18  فولكسفاجن تيجوان  2017  155000       فيراني     بنزين      خصوصي   
19          بيجو 207  2008   23000         خمري     بنزين      خصوصي   

   Car License Transmission Type     Glass  Engine Power  Mileage  \
0     فلسطينية         اوتوماتيك  الكتروني          1600   ١٨٣٠٠٠   
1     فلسطينية         اوتوماتيك  الكتروني          1400    42000   
2     فلسطينية     نصف اوتوماتيك  الكتروني          2000    85000   
3     فلسطينية              عادي  الكتروني          1900    30000   
4     فلسطينية              عادي  الكتروني          1600      NaN   
5     فلسطينية         اوتوماتيك  الكتروني          1200      105   
6     فلسطينية         اوتوماتيك  الكتروني          2200      NaN   
7     فلسطينية              عادي      يدوي          1600      NaN   
8     فلسطينية              عادي  الكتروني          2000      NaN   
9     فلسطينية              عادي      يدوي          1600    50000   
10    فلسطينية         اوتوماتيك  الكتروني          1400   180000   
11    فلسطينية         اوتوماتيك  الكتروني          6700   120000   
12    فلسطينية              عادي  الكتروني          1900  0000014   
13    فلسطينية         اوتوماتيك  الكتروني          1400   220000   
14    فلسطينية              عادي  الكتروني          1900      NaN   
15    فلسطينية              عادي  الكتروني          1200  100 الف   
16    فلسطينية         اوتوماتيك  الكتروني          1400    ١٤٠٠٠   
17    فلسطينية              عادي      يدوي          1600      NaN   
18    فلسطينية         اوتوماتيك  الكتروني          2000    48000   
19    فلسطينية         اوتوماتيك  الكتروني          1600      120   

     Payment Method         Available Previous Owners  \
0   إمكانية التقسيط         للبيع فقط         يد اولى   
1          نقدا فقط  للبيع أو التبديل         يد اولى   
2   إمكانية التقسيط         للبيع فقط        يد ثانيه   
3   إمكانية التقسيط  للبيع أو التبديل               3   
4          نقدا فقط         للبيع فقط             NaN   
5          نقدا فقط         للبيع فقط           ثالثه   
6   إمكانية التقسيط  للبيع أو التبديل             NaN   
7          نقدا فقط         للبيع فقط             NaN   
8   إمكانية التقسيط  للبيع أو التبديل             NaN   
9   إمكانية التقسيط  للبيع أو التبديل               ٣   
10  إمكانية التقسيط         للبيع فقط             NaN   
11  إمكانية التقسيط  للبيع أو التبديل         يد اولى   
12  إمكانية التقسيط  للبيع أو التبديل               ٤   
13  إمكانية التقسيط         للبيع فقط               6   
14         نقدا فقط  للبيع أو التبديل             NaN   
15  إمكانية التقسيط         للبيع فقط        يد ثانية   
16  إمكانية التقسيط  للبيع أو التبديل               ٢   
17         نقدا فقط        

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6864 entries, 0 to 6863
Data columns (total 25 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Make               6864 non-null   object
 1   Year               6864 non-null   int64 
 2   Price              6864 non-null   int64 
 3   Color              6864 non-null   object
 4   Fuel Type          6864 non-null   object
 5   Car Origin         6864 non-null   object
 6   Car License        6864 non-null   object
 7   Transmission Type  6864 non-null   object
 8   Glass              6864 non-null   object
 9   Engine Power       6864 non-null   int64 
 10  Mileage            4923 non-null   object
 11  Payment Method     6864 non-null   object
 12  Available          6864 non-null   object
 13  Previous Owners    5214 non-null   object
 14  Additions          6363 non-null   object
 15  Air Conditioner    6864 non-null   int64 
 16  Central Locking    6864 non-null   int64 


In [7]:
df.isna().sum()

Make                    0
Year                    0
Price                   0
Color                   0
Fuel Type               0
Car Origin              0
Car License             0
Transmission Type       0
Glass                   0
Engine Power            0
Mileage              1941
Payment Method          0
Available               0
Previous Owners      1650
Additions             501
Air Conditioner         0
Central Locking         0
Alarm System            0
CD Player               0
Sunroof                 0
Alloy Wheels            0
Leather Seats           0
Airbag                  0
Passenger Count       777
Drive Type           5433
dtype: int64

- The following columns contain NaN values: 'Mileage', 'Previous Owners', 'Additions', 'Passenger Count' and 'Drive Type'.
- The 'Mileage' column has an 'object' data type; it should be converted to a numeric format.
- For the 'Previous Owners' and 'Passenger Count' columns, similar to the 'Mileage' column, we need to handle noisy data.
- The 'Additions' and 'Drive Type' columns should be dropped.

### We will start working on each column separately to make sure that there's no noisy or missing data

#### Make

In [8]:
#df['Make'].unique()
df['Make'].nunique()

619

split the Make to contain just the brand name

In [9]:
#df['Make'].apply(lambda x: x.split(' ')[0]).unique()
df['Make'] = df['Make'].apply(lambda x: x.split(' ')[0])
df['Make'].nunique()

41

#### Year

In [10]:
df['Year']

0       2008
1       2013
2       2013
3       2010
4       1999
        ... 
6859    1973
6860    2016
6861    2017
6862    2015
6863    2016
Name: Year, Length: 6864, dtype: int64

add age column to compute the age of the car ' will help us later '

In [11]:
Current_Year = 2023
df['age'] = Current_Year - df['Year']
df['age'] = df['age'].astype(np.int64)

Drop year because no need for it now 

In [12]:
df = df.drop('Year', axis=1)

#### Price

In [13]:
df['Price'].unique()

array([ 44000,  58000,  40000,  70000,  33000,  78000,  50000,  18000,
        30000,  60000, 125000,  25000,  32000,  42000,  39500, 155000,
        23000,  68000,  95000, 175000,  65300,  13000,  90000,  85000,
        10000,  36500,  63000,  47000,  48000,  98000, 115000,  27000,
        24000,  35000,  86000,  19000,  39000,  45000,  56000, 100000,
       230000, 110000,  62000,  38000,  75000, 120000,  83000, 109999,
        37000, 350000, 170000,  65000,  80000,  20000,  44999, 113000,
        55000,  26000,   5000,  35500,  97000,  69000,  41000,  89000,
         6000,  72000,  75500, 107000,   9500, 163500,  62500,  49000,
        43500,  43000,  79000,  17000, 117000,  79500, 162000,   2500,
        14000,  12000, 153000,  26500,  92000, 145000,  64000, 138000,
       105000, 270000,  54000,  28500, 135000, 118000,  19500,  59000,
        81000,  57000, 124000,  22000,  99999, 275000,   7000,  48500,
         8500,  94000, 180000,  16000,  46000,  93000, 130000,  61000,
      

#### Color

In [14]:
df['Color'].unique()

array(['أسود', 'فضي', 'أزرق', 'رصاصي', 'أسود ميتالك', 'خمري', 'برتقالي',
       'أحمر', 'أصفر', 'أبيض', 'ذهبي', 'سكني', 'فيراني', 'أبيض عاجي',
       'رمادي', 'أخضر', 'عدة الوان', 'بترولي', 'كحلي', 'زيتي',
       'أزرق سماوي', 'بيج', 'بني', 'كرميدي', 'بنفسجي', 'جيشي'],
      dtype=object)

#### Fuel Type

In [15]:
df['Fuel Type'].unique()

array(['بنزين', 'ديزل', 'هايبرد', 'كهرباء'], dtype=object)

#### Car Origin

In [16]:
df['Car Origin'].unique()

array(['خصوصي', 'عمومي', 'تأجير', 'حكومي', 'تجاري', 'تدريب سياقة'],
      dtype=object)

#### Car License

In [17]:
df['Car License'].unique()

array(['فلسطينية', 'نمرة صفراء'], dtype=object)

#### Transmission Type

In [18]:
df['Transmission Type'].unique()

array(['اوتوماتيك', 'نصف اوتوماتيك', 'عادي'], dtype=object)

#### Glass

In [19]:
df['Glass'].unique()

array(['الكتروني', 'يدوي'], dtype=object)

#### Engine Power

In [20]:
df['Engine Power'].unique()

array([1600, 1400, 2000, 1900, 1200, 2200, 6700, 2500, 1800, 1690, 2800,
       3000, 1500,  999, 1250, 3100, 1000, 2400, 2300, 3600, 3700, 1300,
        230, 2900, 1190,  250, 4000, 3200,  200, 1700, 1150, 6500, 7777,
       6000,  120, 1100,  420, 2700,  903, 3500,  390, 2850, 1995,  800,
        240, 1450, 1009, 1108, 1595,  900,  180, 1605, 1350, 1989, 2600,
       4600, 2150, 6600,  220,  125, 5000, 1495, 1850, 2996, 1999, 2250,
        210, 1235, 1234, 4200, 4800, 1990,  400,  905,  290, 1240, 2005,
       1006, 3300,  280,  140, 4500,  106,  270,  321,  480, 2100, 1498,
        460, 5983, 1518, 2476, 1275, 6400,  300, 2265, 1018, 5500, 5200,
       1950,  260, 2390,  998,  906, 2260,  440, 9000, 2459, 5800, 2006,
       1116,  500, 1242, 4700, 1928, 4400, 4016, 1050, 7000, 1360,  166,
       4750,  150, 2770, 1499, 1840, 1248, 1609,  170, 5985, 1004,  320,
       1444, 3222, 1003, 2008,  160, 1390, 2190,  340, 5555,  249,  550,
        380])

#### Mileage

In [21]:
df['Mileage'].unique()

array(['١٨٣٠٠٠', '42000', '85000', ..., '٣٠٠ الف', '٦٠٠٠', '82الف'],
      dtype=object)

- The data currently have an 'object' data type as we said before and need to be converted to either 'int' or 'float' for numeric operations.
- The data contains Arabic numerals ('١٨٣٠٠٠'), and these need to be translated into their numeric equivalents. 
- The data contains Arabic words ('الف'), which should be converted to '000' or the numbers should be multiplied by 1000.


In [22]:
df['Mileage']

0       ١٨٣٠٠٠
1        42000
2        85000
3        30000
4          NaN
         ...  
6859     50000
6860       140
6861        40
6862    60 الف
6863     82الف
Name: Mileage, Length: 6864, dtype: object

In [23]:
def clean_mileage(mileage):
    try:
        # Remove spaces, translate Arabic numerals to English
        translation_dict = {'٠': '0', '١': '1', '٢': '2', '٣': '3', '٤': '4', '٥': '5', '٦': '6', '٧': '7', '٨': '8', '٩': '9'}
        mileage = ''.join(translation_dict[char] if char in translation_dict else char for char in mileage.replace(' ', ''))

        # Remove Arabic words and commas
        mileage = mileage.replace('الف', '000').replace(',', '')

        # Convert to integer
        mileage = int(mileage)

        return mileage
    except:
        return None

In [24]:
df['Mileage']= df['Mileage'].apply(clean_mileage)

In [25]:
df['Mileage']

0       183000.0
1        42000.0
2        85000.0
3        30000.0
4            NaN
          ...   
6859     50000.0
6860       140.0
6861        40.0
6862     60000.0
6863     82000.0
Name: Mileage, Length: 6864, dtype: float64

#### Payment Method

In [26]:
df['Payment Method'].unique()

array(['إمكانية التقسيط', 'نقدا فقط'], dtype=object)

#### Available

In [27]:
df['Available'].unique()

array(['للبيع فقط', 'للبيع أو التبديل', 'للتبديل فقط'], dtype=object)

#### Previous Owners

In [28]:
df['Previous Owners'].unique()

array(['يد اولى', 'يد ثانيه', '3', nan, 'ثالثه', '٣', '٤', '6',
       'يد ثانية', '٢', 'اولى', '0', 'يد 5', 'يد 4', '11', '2', 'ثانية',
       'يد ثالثة', 'يد 2', 'يد تانيه', '1', 'يد أولى', 'يد٢', 'ايد اولى',
       'يد 0', '5', 'يد ثالثه', '---', '03', 'ايد ثانيه', '000', 'يد٣',
       '،', 'يد أولئ أصحاب سابقون صفر', 'ايد ثانية', '١', 'يد اولي',
       'يد رابعه', 'تانبه', '4', 'غير منمره', 'يد تانية', '٥', 'صفر',
       'أستيراد ألماني', 'يد اولى الماني', '00', '9', 'يد 8', '7',
       'استيراد مباشر من اوروبا', 'يد عاشرة', 'يد خامساً', 'ثالثة',
       'يد اوله', 'اربعة', 'يد صفر', '00000000', 'مستوردة غير منمرة',
       'مش منمره', '10', 'يد تالته', 'مستورد', '000000', '8',
       'مستوردة يد اولى', 'مستوردة غير مرخصة', 'أولى', 'يد 3',
       'استيراد مباشر', '.', 'تانيه', 'يداولى', 'غير منمرمستوردة', 'E',
       '1000', 'صفرر', 'مش منمرة', 'يد خامسه', 'خصوصي', 'خامسة', 'ثانيه',
       'يد ٢', 'رابعه', 'يدثانية', 'يد اةلي', '0000', '٠٠', 'يد أولئ',
       'مش منمر', 'يد3', 'يد ثا

- It have alot of different values for each number 
- We will save all of them on df with its replacement number 
- We will use this df to modify the actual 'Previous Owners' column

In [29]:
def save_lists_to_file():
    no_previous_owners = ['00  ', '0 ', '00 ', 'صفر ', '0000 ', '00000 ', '0 مستورده ', '0مستورد ', 
       'يد صفر  ', 'يد صفر ', '0استيراد  ', '000000 ', 'اصحاب سابقون صفر ', '000 ', 
       'صفر  ', 'صفرض ', '0 مستورد كوري ', '00000000 ', 'يد أولئ ex_owners صفر ', 
       'يد 0 ', 'صفرر ', '، 0 ', ' 0 ', '0 استيراد كوري  ', '0000000 ', ' 00 ', 'مش منمر . (صفر) ', 
       '0 استيراد  ', '٠ ', '٠٠٠٠ ', '٠٠٠٠٠٠٠٠٠ ', '٠٠٠٠٠٠ ', '٠٠٠٠٠٠٠٠٠٠ ', '٠٠ ','٠٠٠٠٠٠٠٠ ' ,
       '٠٠٠٠٠ ', '٠٠٠٠٠٠٠ ', '... ', '...... ', '......... ', 'Zero ', 'لا يوجد ', '. ', 'لايوجد ', 
       '.... ', 'من الشركه مباشر لا يوجد اصحاب سابقون ', 'Ooo ', 'استيراد كوريا - جديد مش منمر ', 
       '  O ',  '.. ', 'لا شيء ', 'O '
       ]
    one_previous_owner = ['1 ', 'يد اولى ', 1, 'يد. اولى', 'واحد ', 'اولا ', '١ ',
       'أولى ', 'يد أولى ', 'اولى  ', 'يد أولى  ', 'يد اولى  ', 'يد أولى استيراد ', 'يد اولي 1 ', 'اولى ', 'يد 1 ', 'يداولى ', 'يد أولى وكالة ', 'يدو اولى ', 'يد اولى شركة ', 
       'يد. اولى ', 'ياولى ', 'يد اولى وكالة ', 'يد اولى  اصل شركة ', 'يد اولى اصحاب سابقون 0 ', 'أولى  ', 'وارد كوريا يد اولى ', 'يد اولى صفر كيلو  ', 'مستورده يد اولى ', 
       'ايد اولى ', '1  ', 'يد أولى بعد الاستيراد ', 'يد أولى 1 ', 'ايد اولى  ', ' 1 ', 'يد اولى فقط ', 'يد أولى،... ', 'اولى\x89 ', 'فقط يد اولى ', 'يد اولى، استيراد شخصي ', 
       'يدو اولى  ', 'يد اولى ( شركة ) ', 'يدأولى ', 'يداولى سحب شركه ', 'يد اولى استيراد شخصي ', 'ايد أولى  ', '01 ', 'يد اولى الماني  ', 'يد اولىً ', 'يد اولى / استيراد ', 
       '1 أولى  ', 'مستوردة يد اولى ', 'بد اولى مستورد ', 'يد اولى من الشركة المتحدة ',
       'يد اولا ', 'يداولا ', 'يد اولا  ', 'يد ١ ', 'يد  ١ ', 'يد واحد  ', '11111 ', 'يد واله ', 'يد اولئ ', 'يد اولي', 'ايد اوله ', 'يد اولة  ', 
       'اوله  ', 'يد اول',  'يدى اوله  ', 'يد اولي  ', 'اول ', 'يد ولا ', 'يد اول  ', 'يد اولي ', 'اوله ', 'يد ولى ', 'يد أولي ', 'يد اول ', 
       'يد أولئ  ', 'يد ا ', 'يد اوله ', 'يد واحدة  ', 'يد ', 'يد اولة ',  '4 نفس الشخص ', 'يد اةلي ', 'أولئ', 'يد أولئ أصحاب سابقون صفر '
       ]
    two_previous_owners = ['2 ', 'يد ثانية ', 'يد ثانيه ', 'ثانية ', '٢ ', 'يد ثانيه  ', 'ثانيه ', 'يد تانيه ', ' يد ثانيه ', 
       'يد ثانية  ', '2  ', 'يد 2 ', 'يدثانيه  ', 'ايد ثانيه ', 'يد ٢ ', ' يد ثانية ', 'يدثانيه ', 'يد تانية ', 
       'ايد 2 ', ' 2 ', 'أيد تانية ', 'يدثانية ', 'تانيه ', '2يد  ', 'يدتانيه ', 'ثانية  ', 'ايد ثانية  ', '2يد ', 
       'يد٢ ', 'يد2 ', 'ايدثانيه ', 'يد  ثانية ', 'يد تانيه  ', 'تانية ', 'الثانية ', 'يد تانية  ', '@ ',
       'اثنان ', 'يد تاني ', 'يد ثني ', 'يد ثاتيه ', 'يد ثاني ', 'ثانبه ', 'يد ثاني.  ', 'تانبه ', 'يد ثانيا ', 
       'تاني  ', 'ثاني ', 'يديانيه ', 'يد ثاني  ',
       ]
    three_previous_owners = ['3 ', 'ثالثة ', 'يد ثالثة ', 'يد ثالثة  ', 'يد ثالثه ', '٣ ', 'يد تالته ', 'يد ثالثة ( فتيات) ', 
              'يد 3 ', 'ثالثه ', '03 ', 'ثالثة  ', '13 ', 'يد ثالثه  ', 'يد3 ', 'يدثالثة ', 'عدد٣ ', 'تالته ', 
              'يد ٣ ', 'يدثالثه ', 'يد٣ ', 'يد التالته ', ' ايد ثالثه ', 'تالثه ', '3  ', ',3 ', 'ايد ثالثة  ', 'ايد 3 ',
              'يد ثالث ', 'ثالث ', 'ثالثى ',  'يد ثالثي ',
              ]
    four_previous_owners = ['4 ', 'رابعه ', '٤ ', 'يد رابعة  ', 'اربعة ', 'يد رابعه ', 'يد 4 ', 'يد رابعة ', '4  ', 'رابعة ', 
              'يد4 ', 'يد ٤ بس نخب السيارة ', 'يد ٤ ', 'يد٤ ', 'يد رابعه  ', 'ايد رابعه ', '14 ', '04 ', 'يد اربعه ', 
              'رابعه  ', '٤٤٤٤٤ ']
    five_previous_owners = ['5 ', 'يد خامسة ', '٥ ', 'يد 5 ', 'يد خامسه ', 'يد خامسة  ', ' 5 ', 'يدخامسه ', 'خامسة ', 
                     'يد٥ ', 'خامسه ', 'يد ٥ ', '  5 ', 'يد خامساً  ', 'يد خمسة ', 'خمسه ', 'خمسة ', 'خمس ', ]
    six_previous_owners = ['يد سادسة ', '6 ', '٦ ', 'يد6 ', '  6 ', 'يد 6 ']
    seven_previous_owners = ['7 ', '٧ ', 'يد سابعة ', 'يد سابعه ', 'يد ٧  ', 'سابعه ']
    eight_previous_owners = ['٨ ', '8 ', '١٨ ', 'يد ثامنه ', 'يد 8 ', 'ثمانيه ']
    nine_previous_owners = ['9 ', '٩ ', '٠٠٠٠٩ ', 'يد ٩  ', 'يد تاسعة ']
    ten_previous_owners = ['10 ', '١٠ ', 'يد عاشره ', 'يد عاشرة ']
    eleven_previous_owners = ['11 ', '١١ ']
    twelve_previous_owners = ['12 ', '١٢ ']
    fourteen_previous_owners = ['١٥ ', '15 ']
    sixteen_previous_owners = ['16 ']
    seventeen_previous_owners = ['17 ', 'يد ١٧ ', '١٧ ', '١٧', '١٧ ', '١٧ ', 'يد ١٧ ','17 ', 'يد ١٧ ' , '١٧ ', 'يد ١٧ ', '١٧ ']
    eighteen_previous_owners = ['18 ']
    nineteen_previous_owners = ['19 ']
    importation = ['مستورد شركة  ', 'استيراد الماني ', ' استيراد الماني ', 'مستورد ', 'استيراد مباشر من كوريا  ', 'استيراد كوري ', 'مستوردة ', 'استيراد ', 'استيراد كوريا ', 'استيراد مباشر من اوروبا  ', 'وارد المانيا ', 'استراد  ', 'استيراد المانيا ', 'استيراد كوريا  ', 'مستورده  ', 'مستوردة  ',  'أستيراد ألماني  ', 'مستوردة غير مرخصة ', 'استراد شخصي ','مستورد  ', 'مستورده ', 'استيراد مباشر ']
    no_number = ['غير منمر استيراد الماني ', 'غير منمرة ', 'غير منمره ', 'مش منمر ', 'مش منمرة  ', 'مش منمرة ', 
                     'غير منمر  ', 'مستود غير منمر ', 'مش منمرا ', 'غير منمرة  ', 'غير منمر ', 'مش منمر  ', 'غير منمرمستوردة ',
                     'مستوردة غير منمرة ', 'مش منمره '
                     ]
    have_number = ['نعم ', ' منمره  ']
    null_values = ['- ', '******** ', 'ا ', '********** ', '****** ', '--- ', '؟ ', '? ', 'E ', '، ', 
                     'للبيع  ', 'اصل تدريب سواقه  ', 'انا ', 'سحب شركة ', 'شطب ', 'اصل شركة ', 'مشطوب ', 'خصوصي  '
                     ]
    large_numbers = ['159000 ', '56 ', '170000 ', '78888877766 ', '0778244 ', '1000 ', '123423 ', '12345 ', 'اشخاص ']


    def save_to_file(replacement_lists, file_path, file_format='csv'):
        data = {'Value': [], 'Replacement': []}

        for replacement_value, lists in replacement_lists.items():
            for value_list in lists:
                for value in value_list:
                    data['Value'].append(value)  # Append values directly without conversion
                    data['Replacement'].append(replacement_value)

        df = pd.DataFrame(data)

        # Handle cases with multiple values in the 'Value' column 
        def handle_multiple_values(row):
            values = row['Value']
            replacement_value = row['Replacement']

            if isinstance(values, int):
                data['Value'].append(values)  # Append value directly without conversion
                data['Replacement'].append(replacement_value)
            else:
                for value in values:
                    data['Value'].append(value)  # Append value directly without conversion
                    data['Replacement'].append(replacement_value)

        df.apply(handle_multiple_values, axis=1)

        df.to_csv(file_path, index=False)


    replacement_lists = {
        0: [no_previous_owners, no_number, have_number, importation],
        1: [one_previous_owner],
        2: [two_previous_owners],
        3: [three_previous_owners],
        4: [four_previous_owners],
        5: [five_previous_owners],
        6: [six_previous_owners],
        7: [seven_previous_owners],
        8: [eight_previous_owners],
        9: [nine_previous_owners],
        10: [ten_previous_owners],
        11: [eleven_previous_owners],
        12: [twelve_previous_owners],
        14: [fourteen_previous_owners],
        16: [sixteen_previous_owners],
        17: [seventeen_previous_owners],
        18: [eighteen_previous_owners],
        19: [nineteen_previous_owners],
        np.NaN: [null_values, large_numbers],
    }
    save_to_file(replacement_lists, '../csv_files/replacement_values.csv', 'csv')

In [30]:
save_lists_to_file()

In [31]:
df['Previous Owners']

0        يد اولى
1        يد اولى
2       يد ثانيه
3              3
4            NaN
          ...   
6859         NaN
6860         NaN
6861           1
6862           2
6863           1
Name: Previous Owners, Length: 6864, dtype: object

In [32]:
replacement_df = pd.read_csv('../csv_files/replacement_values.csv')

In [33]:
replacement_df

Value  Replacement
0        00            0.0
1          0           0.0
2         00           0.0
3        صفر           0.0
4       0000           0.0
..        ...          ...
370  0778244           NaN
371     1000           NaN
372   123423           NaN
373    12345           NaN
374    اشخاص           NaN

[375 rows x 2 columns]

In [34]:
df['Previous Owners'] = df['Previous Owners'].str.strip()
replacement_dict = dict(zip(replacement_df['Value'], replacement_df['Replacement']))
replacement_dict = {key.strip(): value for key, value in replacement_dict.items()}
replacement_series = pd.Series(replacement_dict.values(), index=replacement_dict.keys())
df['Previous Owners'].replace(to_replace=replacement_series.index, value=replacement_series.values, inplace=True)

In [35]:
df['Previous Owners']

0       1.0
1       1.0
2       2.0
3       3.0
4       NaN
       ... 
6859    NaN
6860    NaN
6861    1.0
6862    2.0
6863    1.0
Name: Previous Owners, Length: 6864, dtype: float64

In [36]:
df['Previous Owners'].unique()

array([ 1.,  2.,  3., nan,  4.,  6.,  0.,  5., 11.,  9.,  8.,  7., 10.,
       17., 12., 14., 18., 19., 16.])

done ")

#### Additions

Drop year because no need for it now since each addition feature have separate column

In [37]:
df = df.drop('Additions', axis=1)

#### Binary features 
- Air Conditioner
- Central Locking
- Alarm System
- CD Player
- Sunroof
- Alloy Wheels
- Leather Seats
- Airbag

In [38]:
print(df['Air Conditioner'].unique())
print(df['Central Locking'].unique())
print(df['Alarm System'].unique())
print(df['CD Player'].unique())
print(df['Sunroof'].unique())
print(df['Alloy Wheels'].unique())
print(df['Leather Seats'].unique())
print(df['Airbag'].unique())

[1 0]
[1 0]
[1 0]
[1 0]
[0 1]
[0 1]
[1 0]
[1 0]


There's no Noisy Data 

#### Passenger Count

In [39]:
#df['Passenger Count'].nunique()
df['Passenger Count'].unique()

array([nan, '4+1', '5+1', '4+١', '2', '8+1', '1+2', '6+1', '7+1', '3+1',
       '2+1', '1+1', '٤+١', 'اكثر من 10', '٧+١', '٧', '6x1', '1', '9+1',
       '1+7', '٧+1', '6+١', '7.1', '١+٤', '8', '٦+١', '7,+ 1', '000',
       '6±١', '+1', '7 =1', 'مثال: 4 + 1', '٨+١', '1٠4', '22', '7:1',
       '4\\١', '1+٤', '4+2', '٥+١', '5+2', '2/1'], dtype=object)

In [40]:
df['Passenger Count'].replace({'اكثر من 10': np.nan, 'مثال: 4 \+ 1': np.nan, '000': np.nan}, inplace=True, regex=True)
df['Passenger Count'] = df['Passenger Count'].str.replace('±', '+')
df['Passenger Count'].replace({'6x1': '6+1', '7.1': '7+1', '7 =1': '7+1', '7:1': '7+1', '4\\\\١': '4+1','2/1':'2'}, inplace=True, regex=True)
df['Passenger Count'] = df['Passenger Count'].replace('[^\d+]', '', regex=True)
arabic_numerals = {'١': '1', '٤': '4', '٥': '5', '٦': '6', '٧': '7', '٨': '8', '٩': '9' , '٠':'0'}
for arabic, numeral in arabic_numerals.items():
    df['Passenger Count'] = df['Passenger Count'].str.replace(arabic, numeral)

In [41]:
#df['Passenger Count'].nunique()
df['Passenger Count'].unique()

array([nan, '4+1', '5+1', '2', '8+1', '1+2', '6+1', '7+1', '3+1', '2+1',
       '1+1', '7', '1', '9+1', '1+7', '1+4', '8', '+1', '104', '22',
       '4+2', '5+2'], dtype=object)

Convert Strings '+' to Numeric Values

In [42]:
df['Passenger Count'] = df['Passenger Count'].str.extract('(\d+)').astype(float)

In [43]:
df['Passenger Count'].unique()

array([ nan,   4.,   5.,   2.,   8.,   1.,   6.,   7.,   3.,   9., 104.,
        22.])

#### Drive Type

There's alot of nan values '5433' so we will drop it 

In [44]:
df = df.drop('Drive Type', axis=1)

In [45]:
df

Make   Price  Color Fuel Type Car Origin Car License  \
0        كيا   44000   أسود     بنزين      خصوصي    فلسطينية   
1     هونداي   58000    فضي     بنزين      خصوصي    فلسطينية   
2       بيجو   40000   أسود     بنزين      خصوصي    فلسطينية   
3      سكودا   70000   أزرق      ديزل      خصوصي    فلسطينية   
4     سوبارو   33000  رصاصي     بنزين      خصوصي    فلسطينية   
...      ...     ...    ...       ...        ...         ...   
6859    فيات   27000   أحمر     بنزين      خصوصي    فلسطينية   
6860  هونداي  117000   أسود      ديزل      خصوصي    فلسطينية   
6861  هونداي   67000    فضي     بنزين      خصوصي    فلسطينية   
6862   سكودا   58000   أبيض     بنزين      خصوصي    فلسطينية   
6863     سيت   65000   سكني     بنزين      خصوصي    فلسطينية   

     Transmission Type     Glass  Engine Power   Mileage   Payment Method  \
0            اوتوماتيك  الكتروني          1600  183000.0  إمكانية التقسيط   
1            اوتوماتيك  الكتروني          1400   42000.0         نقدا فقط   
2        نصف اوتوماتيك  الكتروني          2000   85000.0  إمكانية التقسيط   
3                 عادي  الكتروني          1900   30000.0  إمكانية التقسيط   
4                 عادي  الكتروني          1600       NaN         نقدا فقط   
...                ...       ...           ...       ...              ...   
6859              عادي      يدوي          1800   50000.0         نقدا فقط   
6860         اوتوماتيك  الكتروني          2000     140.0         نقدا فقط   
6861         اوتوماتيك  الكتروني          1400      40.0         نقدا فقط   
6862         اوتوماتيك  الكتروني          1200   60000.0         نقدا فقط   
6863              عادي  الكتروني          1400   82000.0  إمكانية التقسيط   

             Available  Previous Owners  Air Conditioner  Central Locking  \
0            للبيع فقط              1.0                1                1   
1     للبيع أو التبديل              1.0                1                1   
2            للبيع فقط              2.0                1                1   
3     للبيع أو التبديل              3.0                1                1   
4            للبيع فقط              NaN                1                1   
...                ...              ...              ...              ...   
6859         للبيع فقط              NaN                0                0   
6860         للبيع فقط              NaN                1                1   
6861         للبيع فقط              1.0                1                1   
6862  للبيع أو التبديل              2.0                1                1   
6863         للبيع فقط              1.0                1                1   

      Alarm System  CD Player  Sunroof  Alloy Wheels  Leather Seats  Airbag  \
0                1          1        0             0              1       1   
1                1          0        0             1              1       1   
2                1          1        1             1              1       1   
3                1          1        0             1              1       1   
4                1          1        0             1              0       1   
...            ...        ...      ...           ...            ...     ...   
6859             0          1        0             0              1       0   
6860             1          0        1             1              1       1   
6861             1          1        0             1              1       1   
6862             1          1        0             1              0       1   
6863             1          0        0             1              1       1   

      Passenger Count  age  
0                 NaN   15  
1                 4.0   10  
2                 4.0   10  
3                 4.0   13  
4                 4.0   24  
...               ...  ...  
6859              NaN   50  
6860              6.0    7  
6861              4.0    6  
6862              4.0    8  
6863              NaN    7  

[6864 rows x 23 columns]

### Work on Missing data 

In [46]:
df[df.columns[df.isna().any()]].isna().sum()

Mileage            2119
Previous Owners    1682
Passenger Count     788
dtype: int64

In [47]:
df.dropna(subset=['Mileage', 'Previous Owners', 'Passenger Count'], how='all', inplace=True)

In [48]:
df[df.columns[df.isna().any()]].isna().sum()

Mileage            1782
Previous Owners    1345
Passenger Count     451
dtype: int64

thats make it much better

we will continue on next part after detecting outliers and handling them so they won't effect the data

In [49]:
df.to_csv('../csv_files/car_df.csv', index=False)